File: mini-agent/internal/rag/chunk.go
// Package rag 实现 RAG（Retrieval-Augmented Generation）链路中"文档侧"的能力。
//
// 在 agent 链路中的位置：RAG = 文档 → 切块（chunking）→ embedding 入库 →
// 检索 top-k → 把相关块组装进 prompt 交给 LLM。整条链路横跨多个包：
//   - 本包（internal/rag）：切块（本文件，练习 3）与检索工具 kb_search（练习 4）；
//   - internal/embed（练习 1）：把文本翻译成向量；
//   - internal/vectorstore（练习 2）：存向量并做相似度检索。
//
// 为什么切块单独成章：LLM 的上下文窗口和 embedding 模型的输入长度都有限，
// 文档不能整篇塞进 prompt，也不能整篇只 embedding 成一个向量（主题会被稀释）。
// 切块质量直接决定检索质量的上限——检索回来的块如果断章取义或充满噪声，
// 后面的 prompt 工程做得再好也救不回来。
package rag

import (
	"strings"
)

// ChunkOptions 是文档切块的调参入口。
//
// MaxChars 是 RAG 的第一调参位（面试高频：chunk 切多大？）：
//   - 太大 → 一个块里混入多个主题，embedding 向量被"平均"稀释，检索命中率下降；
//   - 太小 → 上下文不完整，答案断章取义（"第 3 条"被切走，块里只剩"……除外"）。
//
// OverlapChars 是相邻块之间的重叠长度，防止语义在块边界被拦腰切断：
// 一句话横跨两个块时，重叠保证它在至少一个块里是完整的。
// 常见经验值是块大小的 10%~20%。
type ChunkOptions struct {
	MaxChars     int // 每个块的最大字符数（rune 计）
	OverlapChars int // 相邻硬切块之间的重叠字符数（rune 计）
}

// DefaultChunkOptions 返回教学用默认值：400 字符块 + 60 字符重叠。
//
// 注意：这里按字符（rune）而不是 token 计数，是有意的简化——
// 教学项目不引入 tokenizer 库（如 tiktoken），token 精确计数需要
// 加载模型对应的词表文件，成本高且与"先手写原理"的约定相悖。
// 面试可主动讲 token 与字符的换算经验值：
// 中文 1 token ≈ 1.5~2 字符，英文 1 token ≈ 4 字符。
// 因此 400 字符 ≈ 中文 200~270 token / 英文 100 token，
// 落在常见起点（200~500 token）的偏保守一侧，对 bge-m3
// （最大输入 8192 token）非常安全。
func DefaultChunkOptions() ChunkOptions {
	return ChunkOptions{MaxChars: 400, OverlapChars: 60}
}

// TODO(练习3): chunking 文档切分
//
// 【任务】
// 实现 func Chunk(text string, opts ChunkOptions) []string：
// 把一篇文档切成若干不超过 opts.MaxChars 字符（rune）的块，供后续
// embedding 入库。签名已给出，你只需要填实现。
//
// 【提示】
//   - 策略：结构优先。先按空行把文本切成段落（strings.Split(text, "\n\n")，
//     每段 TrimSpace、丢弃空段），再贪心地把段落依次打包进一个块：
//     当前块放得下（连同段落间分隔符 "\n\n" 一起计）就追加，放不下就
//     封存当前块、开新块。段落不拆——结构完整的块检索质量最好。
//   - 单段自身超过 MaxChars 时，对该段硬切（固定窗口）：从 rune 偏移
//     start 处取 MaxChars 个 rune 为一块，下一块从 start + MaxChars -
//     OverlapChars 处开始，即相邻块重叠 OverlapChars 个 rune。
//   - 最大的坑：必须用 []rune 切片计量和切分，绝不能用 byte 下标——
//     中文一个字符占 3 个 byte，按 byte 切会把一个汉字从中间劈开，
//     产出乱码（面试常考的 UTF-8 考点）。
//   - 防御 opts 异常值：MaxChars <= 0 时用 DefaultChunkOptions() 兜底；
//     OverlapChars >= MaxChars 时硬切的步长（MaxChars - OverlapChars）
//     会变成 0 或负数导致死循环，必须把 overlap 钳制到 [0, MaxChars-1]。
//   - 不产空块（每块 TrimSpace 后为空则丢弃）；尽量保证原文内容全覆盖
//     （除被 TrimSpace 掉的空白外，任何一段文字都应出现在至少一个块里）。
//
// 【验收】
// go test ./internal/rag/ 通过（测试由参考答案提供，需覆盖：多段落打包
// 不超限不拆段、超长段硬切且重叠正确、纯中文无乱码、空输入返回 nil、
// OverlapChars >= MaxChars 不死循环）。
//
// 参考答案：docs/solutions/stage-02/exercise-3-chunking.md（完成后再看）
func Chunk(text string, opts ChunkOptions) []string {
	opts = normalizeChunkOptions(opts)

	var chunks []string
	var cur []string
	curLen := 0

	flush := func() {
		if len(cur) == 0 {
			return
		}

		chunks = append(chunks, strings.Join(cur, "\n\n"))
		cur = nil
		curLen = 0
	}

	for _, para := range splitParagraphs(text) {
		paraLen := len([]rune(para))
		if len(para) > opts.MaxChars {
			flush()
			chunks = append(chunks, hardCut(para, opts)...)
			continue
		}
		addLen := paraLen
		if len(cur) > 0 {
			addLen += 2
		}

		if curLen+addLen > opts.MaxChars {
			flush()
		}

		if len(cur) == 0 {
			curLen = paraLen
		} else {
			curLen += 2 + paraLen
		}

		cur = append(cur, para)

	}
	flush()
	return chunks
}

func normalizeChunkOptions(opts ChunkOptions) ChunkOptions {
	if opts.MaxChars <= 0 {
		opts = DefaultChunkOptions()
	}

	if opts.OverlapChars < 0 {
		opts.OverlapChars = 0
	}

	if opts.OverlapChars >= opts.MaxChars {
		opts.OverlapChars = opts.MaxChars - 1
	}
	return opts
}

func splitParagraphs(text string) []string {
	var paras []string

	for _, p := range strings.Split(text, "\n\n") {
		if p = strings.TrimSpace(p); p != "" {
			paras = append(paras, p)
		}
	}
	return paras
}

func hardCut(para string, opts ChunkOptions) []string {
	runes := []rune(para)
	step := opts.MaxChars - opts.OverlapChars
	var out []string

	for start := 0; start < len(runes); start += step {
		end := start + opts.MaxChars

		if end > len(runes) {
			end = len(runes)
		}
		out = append(out, string(runes[start:end]))
		if end == len(runes) {
			break
		}
	}
	return out
}

File: stage-02-kb-agent/lib/chunk.ts
/**
 * lib/chunk.ts —— 文档切分（chunking）：把长文档切成适合 embedding 的块。
 *
 * 在 RAG 链路中的位置：文档解析之后、embedding 之前。
 * chunk 是"RAG 的第一调参位"：块太大 → 噪声多、相似度被稀释；
 * 块太小 → 上下文不完整、答案断章取义。
 *
 * 本模块与 Go 侧 mini-agent/internal/rag/chunk.go（练习 3）策略一致，
 * 做练习时建议先回顾 Go 侧练习 3 的参考答案
 * （docs/solutions/stage-02/exercise-3-chunking.md），本练习是它的 TS 移植。
 */

/** ChunkOptions 控制切分粒度。 */
export interface ChunkOptions {
  /** 每块最大字符数（按码点计，见 chunk 的 TODO 提示）。 */
  maxChars: number;
  /** 硬切时相邻块的重叠字符数，防上下文在边界被切断。 */
  overlapChars: number;
}

/**
 * DEFAULT_CHUNK_OPTIONS 经验起点：400 字符 + 15% 重叠。
 * 按字符而非 token 计数是简化（中文 1 字 ≈ 1-2 token，英文 1 词 ≈ 1-2 token），
 * 学习项目够用；生产环境会按 tokenizer 精确计数。
 */
export const DEFAULT_CHUNK_OPTIONS: ChunkOptions = {
  maxChars: 400,
  overlapChars: 60,
};

// TODO(练习6): 文档切分 chunk
//
// 【任务】实现：
//
//   export function chunk(text: string, options?: Partial<ChunkOptions>): string[]
//
// 把一篇文档切成若干不超过 maxChars 字符的块。策略与 Go 侧练习 3 完全一致——
// "结构优先，窗口兜底"：
//  1. 先按空行（"\n\n"）切段落，TrimSpace 后丢弃空段，再贪心地把整段
//     打包进块（段落不被拆散；段落间 "\n\n" 分隔符也占块内额度）；
//  2. 单段自身超过 maxChars 时，先封存当前块，再对该段做固定窗口硬切
//     （步长 = maxChars - overlapChars）。
// 防御：options 缺省时用 DEFAULT_CHUNK_OPTIONS 补齐；overlapChars 必须钳制到
// [0, maxChars-1]，否则步长为 0 或负数，硬切循环永不前进（死循环）。
//
// 【提示】
//   - 本练习最核心的坑：JS 字符串按 UTF-16 code unit 索引，不是按"字符"！
//     str.length、str.slice()、str[i] 都以 code unit 为单位。
//     中文等 BMP 字符恰好 1 个 code unit，看起来安全；
//     但 emoji 和很多生僻字是代理对（2 个 code unit），
//     按 code unit 切会把一个 emoji 劈成两个非法的"半个字符"（乱码）。
//     做法：切分和计量都基于码点数组 —— const chars = Array.from(text)
//     （或 [...text]，for..of 同理），chars.length 才是"字符数"，
//     切完用 chars.slice(start, end).join("") 拼回字符串。
//     这与 Go 侧"按 rune 不按 byte"是同一个坑在不同语言的形态。
//   - 结构参考 Go 侧练习 3 答案的三个函数：normalize（参数钳制）、
//     splitParagraphs（按空行切段）、hardCut（固定窗口硬切）+
//     主函数贪心打包。TS 版可以完全同构。
//   - 覆盖性不变量：除被 trim 掉的空白外，原文任何一段文字都应出现在
//     至少一个块里。自检方法：硬切块去重叠拼接后应还原原段落。
//
// 【验收】启动 dev server 后用 curl 上传样例文档（见
// app/api/ingest/route.ts 的 TODO 验收），返回的块数符合预期；
// 无 key 时配合 EMBEDDING_MOCK=1 跑通全链路。

const codePoints = (s: string): string[] => Array.from(s);

// 参考答案：docs/solutions/stage-02/exercise-6-ingest-pipeline.md（完成后再看）
export function chunk(text: string, options?: Partial<ChunkOptions>): string[] {
  const opts: ChunkOptions = normalizeChunkOptions(options);

  const chunk: string[] = [];
  let cur: string[] = [];
  let curLen = 0;

  const flush = () => {
    if (cur.length == 0) return;

    chunk.push(cur.join("\n\n"));
    cur = [];
    curLen = 0;
  };

  for (const para of splitParagraphs(text)) {
    const paraLen = codePoints(para).length;
    if (paraLen > opts.maxChars) {
      flush();
      chunk.push(...hardCut(para, opts));
      continue;
    }
    const addLen = paraLen + (cur.length > 0 ? 2 : 0);
    if (curLen + addLen > opts.maxChars) {
      flush();
    }

    curLen = cur.length === 0 ? paraLen : curLen + 2 + paraLen;
    cur.push(para);
  }
  flush();
  return chunk;
}

function splitParagraphs(text: string): string[] {
  return text
    .split("\n\n")
    .map((p) => p.trim())
    .filter((p) => p !== "");
}

function hardCut(para: string, opts: ChunkOptions): string[] {
  const chars = codePoints(para);
  const step = opts.maxChars - opts.overlapChars;
  const out: string[] = [];
  for (let start = 0; start < chars.length; start += step) {
    const end = Math.min(start + opts.maxChars, chars.length);
    out.push(chars.slice(start, end).join(""));
    if (end === chars.length) {
      break;
    }
  }
  return out;
}

function normalizeChunkOptions(options?: Partial<ChunkOptions>): ChunkOptions {
  const opts: ChunkOptions = { ...DEFAULT_CHUNK_OPTIONS, ...options };
  if (opts.maxChars <= 0) {
    opts.maxChars = DEFAULT_CHUNK_OPTIONS.maxChars;
  }

  if (opts.overlapChars < 0) {
    opts.overlapChars = 0;
  }

  if (opts.overlapChars >= opts.maxChars) {
    opts.overlapChars = opts.maxChars - 1;
  }
  return opts;
}
